### PR TITLE
[RISCV] Add lowerVECTOR_SHUFFLEAsCONCAT_VECTORS.

### PR DESCRIPTION
**Pull request by HanKuanChen**

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+; CHECK-LABEL: test1:
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    vmv1r.v v9, v8
+; CHECK-NEXT:    vsetivli zero, 16, e32, m1, ta, ma
+; CHECK-NEXT:    vslideup.vi v9, v8, 8
+; CHECK-NEXT:    vmv.v.v v8, v9
+; CHECK-NEXT:    ret
+entry:
+  %1 = call <16 x float> @llvm.vector.insert.v16f32.v8f32(<16 x float> poison, <8 x float> poison, i64 8)
+  %2 = call <16 x float> @llvm.vector.insert.v16f32.v8f32(<16 x float> %1, <8 x float> %0, i64 0)
```
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll:14]: Does inserting a poison subvector into a posion vector do anything?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -5130,6 +5130,40 @@ static SDValue lowerVECTOR_SHUFFLEAsRotate(ShuffleVectorSDNode *SVN,
...
+  assert(SVN->getOperand(1).isUndef());
+  SDValue V1 = SVN->getOperand(0);
+  if (V1.getOpcode() != ISD::INSERT_SUBVECTOR)
+    return SDValue();
+  if (!V1.getOperand(0).isUndef())
+    return SDValue();
+  SDValue InsertValue = V1.getOperand(1);
+  MVT SubVecVT = InsertValue.getSimpleValueType();
+  unsigned SubVecNumElements = SubVecVT.getVectorNumElements();
+  uint64_t InsertIndex = cast<ConstantSDNode>(V1.getOperand(2))->getZExtValue();
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:5150]: You could simplify the zero check and maybe move it beside the undef check
```c++
  if (!V1.getOperand(0).isUndef() || V1.getConstantOperandVal(2) != 0)
    return SDValue();
```

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll (hunk 1 above)
Comment (HanKuanChen) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll]: Actually it does nothing. Remove it already.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (HanKuanChen) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: done

**Review by lukel97 — APPROVED** [review]
The PR could use a small description of the transform but otherwise LGTM

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll [hunk 3]
```diff
@@ -0,0 +1,32 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple riscv64 -mattr=+v -riscv-v-vector-bits-min=512 -verify-machineinstrs < %s | FileCheck %s
+
+define <16 x float> @test1(<8 x float> %0) {
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll:4]: InstCombine already flattens both of these tests. Do you have other examples?

```
define <16 x float> @test1(<8 x float> %0) {
entry:
  %1 = shufflevector <8 x float> %0, <8 x float> poison, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  ret <16 x float> %1
}

define <16 x i32> @test2(<4 x i32> %0) {
entry:
  %1 = shufflevector <4 x i32> %0, <4 x i32> poison, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3>
  ret <16 x i32> %1
}define <16 x float> @test1(<8 x float> %0) {
entry:
  %1 = shufflevector <8 x float> %0, <8 x float> poison, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  ret <16 x float> %1
}

define <16 x i32> @test2(<4 x i32> %0) {
entry:
  %1 = shufflevector <4 x i32> %0, <4 x i32> poison, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3, i32 0, i32 1, i32 2, i32 3>
  ret <16 x i32> %1
}
```

Comment (HanKuanChen) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll:4]: That is true. But I want to make sure INSERT_SUBVECTOR + VECTOR_SHUFFLE can be folded into CONCAT_VECTORS. SLP will generate the code like this.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll:4]: InstCombine should run after SLP. Do you have examples where InstCombine isn't cleaning it up? If you do, we should investigate InstCombine first.

Comment (HanKuanChen) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll:4]: No for now.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll:4]: This also doesn't seem like a RISC-V specific optimization.

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-concat-vectors.ll:4]: I was thinking that as well. It looks like GISel canonicalizes shuffles to concats: https://reviews.llvm.org/D69149

I didn't notice any codegen changes with this patch when building SPEC or the test suite so I presume that's InstCombine already taking care of this.

**Comment by HanKuanChen** [comment]
Close it since instcombine can solve this.

**Pull request closed without merging**